### PR TITLE
Omit unique module versions from canonical repo names

### LINKED ISSUE #20997: Usage of bzlmod causes cache invalidations due to module version being in file path
### Description of the feature request:

Remove module version from the visible module path as it causes frequent cache invalidations when modules updates. This is particularly problematic in C++ where changes in include paths due to module versions can propagate a cache invalidation over most of the build graph; despite no actual dependent files changing.

### Which category does this issue belong to?

External Dependency

### What underlying problem are you trying to solve with this feature?

_No response_

### Which operating system are you running Bazel on?

Mac OS

### What is the output of `bazel info release`?

release 6.4.0

### If `bazel info release` returns `development version` or `(@non-git)`, tell us how you built Bazel.

_No response_

### What's the output of `git remote get-url origin; git rev-parse master; git rev-parse HEAD` ?

_No response_

### Have you found anything relevant by searching the web?

_No response_

### Any other information, logs, or outputs that you want to share?

_No response_

### PR DESCRIPTION
**Pull request by fmeum**
The canonical repository name of a module `rules_foo` is now `rules_foo~` instead of e.g. `rules_foo~1.2.3` if there is only a single version of the module in the entire dep graph. This also applies to overrides, which previously used the canonical repository name `rules_foo~override`.

This improves cacheability of actions across module version changes and also prevents the output base from filling up with outdated module and extension repositories. See the long comment in `ModuleKey#getCanonicalRepoName` for a detailed explanation of why this particular scheme was chosen.

The change includes a bump of the lockfile version as the canonical repo names of label attributes in extension usages are affected.

RELNOTES: The scheme for generating canonical repository names has changed to improve cacheability of actions across dependency version updates. Note that canonical names are not considered to be public API and can change at any time. See https://bazel.build/external/module#repository_names_and_strict_deps for advice on how to avoid hardcoding canonical repository names.

Fixes #20997

### CODE REVIEW

**Comment by fmeum** [comment]
@Wyverald @meteorcloudy Could you already take a look at the implementation? If you agree with the new naming scheme and the overall implementation, I would go ahead and fix all the tests.

File: src/main/java/com/google/devtools/build/lib/bazel/bzlmod/BazelDepGraphFunction.java [hunk 1]
```diff
@@ -249,6 +264,27 @@ static BzlmodFlagsAndEnvVars getFlagsAndEnvVars(Environment env) throws Interrup
     return extensionUsagesTableBuilder.buildOrThrow();
   }
 
+  private static ImmutableBiMap<RepositoryName, ModuleKey> computeCanonicalRepoNameLookup(
+      ImmutableMap<ModuleKey, Module> depGraph) {
+    // Find those modules of which there is only a single version in the dep graph. Currently, the
```
Comment (Wyverald) [inline src/main/java/com/google/devtools/build/lib/bazel/bzlmod/BazelDepGraphFunction.java:269]: s/those modules of which there is/modules with/

may make it a bit easier to read

File: src/main/java/com/google/devtools/build/lib/bazel/bzlmod/BazelDepGraphFunction.java [hunk 2]
```diff
@@ -249,6 +264,27 @@ static BzlmodFlagsAndEnvVars getFlagsAndEnvVars(Environment env) throws Interrup
...
+    // only way to have multiple versions of a module in the dep graph is via
+    // multiple_version_override.
+    ImmutableSet<String> uniqueVersionModules =
+        depGraph.keySet().stream()
+            .collect(groupingBy(ModuleKey::getName, counting()))
```
Comment (Wyverald) [inline src/main/java/com/google/devtools/build/lib/bazel/bzlmod/BazelDepGraphFunction.java:274]: could pass in the overrides and filter out multiple_version_overrides instead, WDYT? I think that's slightly more intuitive (and realisitically I don't think we'll ever introduce other ways to have multiple versions of a module)

File: src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleKey.java [hunk 3]
```diff
@@ -83,15 +75,45 @@ public final String toString() {
...
+      // * By making it so that the module name and the canonical repository name of a module are
+      //   never identical, even when using an override, we introduce "grease" that intentionally
+      //   tickles bugs in code that doesn't properly distinguish between the two, e.g., by not
+      //   applying repo mappings. Otherwise, these bugs could go unnoticed in BCR test modules and
+      //   would only be discovered when used with a `multiple_version_override`, which is very
+      //   rarely used.
+      suffix = "";
+    } else {
+      // getVersion().isEmpty() is true only for modules with non-registry overrides, which imply
+      // enforce there is a single version of the module in the dep graph.
```
Comment (Wyverald) [inline src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleKey.java:112]: remove one of "imply enforce"

File: src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleFileFunction.java [hunk 4]
```diff
@@ -359,7 +366,9 @@ private GetModuleFileResult getModuleFile(
     // If there is a non-registry override for this module, we need to fetch the corresponding repo
     // first and read the module file from there.
     if (override instanceof NonRegistryOverride) {
-      RepositoryName canonicalRepoName = key.getCanonicalRepoName();
+      // A module with a non-registry override always has a unique version across the entire dep
+      // graph.
+      RepositoryName canonicalRepoName = key.getCanonicalRepoName(/* hasUniqueVersion */ true);
```
Comment (Wyverald) [inline src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleFileFunction.java:371]: hasUniqueVersion=

File: src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleKey.java [hunk 5]
```diff
@@ -83,15 +75,45 @@ public final String toString() {
   }
 
   /** Returns the canonical name of the repo backing this module. */
-  public RepositoryName getCanonicalRepoName() {
+  public RepositoryName getCanonicalRepoName(boolean hasUniqueVersion) {
```
Comment (Wyverald) [inline src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleKey.java:78]: I wonder if we can come up with separate names for the two concepts so we don't need to have an awkward boolean parameter. `getFullCanonicalRepoName` & `getClippedCanonicalRepoName`, for example? (and we just say "for modules with a single version, use the clipped canonical repo name"). or is that even a good idea?

**Review by Wyverald — COMMENTED** [review]
overall LGTM

File: src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleKey.java [hunk 6]
```diff
@@ -83,15 +75,45 @@ public final String toString() {
...
       return WELL_KNOWN_MODULES.get(getName());
     }
     if (ROOT.equals(this)) {
       return RepositoryName.MAIN;
     }
-    return RepositoryName.createUnvalidated(
-        String.format("%s~%s", getName(), getVersion().isEmpty() ? "override" : getVersion()));
+    String suffix;
+    if (hasUniqueVersion) {
+      // If there is a unique version of this module in the entire dep graph, we elide the version
```
Comment (meteorcloudy) [inline src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleKey.java:87]: Can you please update the doc here as well: https://bazel.build/external/module#repository_names_and_strict_deps?

File: src/test/py/bazel/bzlmod/bazel_lockfile_test.py [hunk 7]
```diff
@@ -1741,19 +1742,19 @@ def testExtensionRepoMappingChange_loadsAndRepoRelativeLabels(self):
...
+            'bazel_dep(name="bar",version="1.0",repo_name="repo_name")',
             'ext = use_extension(":ext.bzl", "ext")',
             'use_repo(ext, "repo")',
         ],
     )
     _, _, stderr = self.RunBazel(['build', ':lol'])
-    self.assertIn('STR=@@foo~2.0//:BUILD', '\n'.join(stderr))
+    self.assertIn('STR=@@bar~//:BUILD', '\n'.join(stderr))
 
   def testExtensionRepoMappingChange_sourceRepoNoLongerExistent(self):
```
Comment (fmeum) [inline src/test/py/bazel/bzlmod/bazel_lockfile_test.py:1759]: @Wyverald Do you happen to have an idea how to port this test over to the new naming scheme?

File: src/main/java/com/google/devtools/build/lib/bazel/bzlmod/BazelDepGraphFunction.java (hunk 2 above)
Comment (Wyverald) [reply src/main/java/com/google/devtools/build/lib/bazel/bzlmod/BazelDepGraphFunction.java:272]: `multiple_version_override` doesn't allow less than 2 versions. But anyway, I don't feel too strongly about this, and there's pros and cons either way, so I'm fine with keeping it as is.

File: src/test/py/bazel/bzlmod/bazel_lockfile_test.py (hunk 7 above)
Comment (Wyverald) [reply src/test/py/bazel/bzlmod/bazel_lockfile_test.py:1759]: worst case scenario, we could manually alter the recorded repo mapping entries to contain a nonexistent source repo. that would probably simplify the test setup a lot too.

**Comment by fmeum** [comment]
I fixed all tests except for the distfile setup. I won't make changes to anything other than the bootstrap tests and lockfile patching logic, which I haven't been able to adapt yet.

**Comment by fmeum** [comment]
@Wyverald @meteorcloudy Tests are passing now, I had to update the canonical name of `r_j_e`.

File: site/en/external/module.md [hunk 8]
```diff
@@ -190,19 +190,34 @@ flexibility.
...
 The [apparent name](/external/overview#apparent-repo-name) of a repo backing a
 module to its direct dependents defaults to its module name, unless the
 `repo_name` attribute of the [`bazel_dep`](/rules/lib/globals/module#bazel_dep)
 directive says otherwise. Note that this means a module can only find its direct
 dependencies. This helps prevent accidental breakages due to changes in
 transitive dependencies.
 
+The [canonical name](/external/overview#canonical-repo-name) of a repo backing a
+module is `{{ "<var>" }}module_name{{ "</var>" }}~{{ "<var>" }}version{{
+"</var>" }}` (for example, `bazel_skylib~1.0.3`). For modules with a unique
```
Comment (meteorcloudy) [inline site/en/external/module.md:202]: The first sentence is not correct anymore, right?

**Review by meteorcloudy — APPROVED** [review]
Thank you! ❤️

**Comment by fmeum** [comment]
@meteorcloudy I addressed your comment and also added RELNOTES to the PR description.

**Review by Wyverald — APPROVED** [review]

**Comment by fmeum** [comment]
@bazel-io fork 7.1.0

**Comment by iancha1992** [comment]
The changes in this PR have been included in [Bazel 7.1.0 RC1](https://releases.bazel.build/7.1.0/rc1/index.html). Please test out the release candidate and report any issues as soon as possible. If you're using Bazelisk, you can point to the latest RC by setting USE_BAZEL_VERSION=last_rc.
Thanks!

**Pull request closed without merging**